File: prerequisite_service/tags/version-1.0.0.1/src/java/main/com/topcoder/service/prerequisite/PrerequisiteDocument.java
/*
 * Copyright (C) 2008 TopCoder Inc., All Rights Reserved.
 */
package com.topcoder.service.prerequisite;

import java.io.Serializable;

import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorType;
import javax.xml.bind.annotation.XmlElement;
import javax.xml.bind.annotation.XmlType;
import javax.xml.datatype.XMLGregorianCalendar;

/**
 * <p>
 * It represents the pre-requisite document entity (bean) used by the service. It contains the information of a
 * pre-requisite document. The information can be null or empty, therefore this check is not present in the setters. It
 * is similar and related to <code>DocumentVersion</code> if prerequisite manager.
 * </p>
 * <p>
 * The following schema fragment specifies the expected content contained within this class.
 * </p>
 *
 * <pre>
 * &lt;complexType name=&quot;prerequisiteDocument&quot;&gt;
 *   &lt;complexContent&gt;
 *     &lt;restriction base=&quot;{http://www.w3.org/2001/XMLSchema}anyType&quot;&gt;
 *       &lt;sequence&gt;
 *         &lt;element name=&quot;documentId&quot; type=&quot;{http://www.w3.org/2001/XMLSchema}long&quot;/&gt;
 *         &lt;element name=&quot;version&quot; type=&quot;{http://www.w3.org/2001/XMLSchema}Integer&quot;/&gt;
 *         &lt;element name=&quot;versionDate&quot; type=&quot;{http://www.w3.org/2001/XMLSchema}date&quot;/&gt;
 *         &lt;element name=&quot;name&quot; type=&quot;{http://www.w3.org/2001/XMLSchema}string&quot;/&gt;
 *         &lt;element name=&quot;content&quot; type=&quot;{http://www.w3.org/2001/XMLSchema}string&quot;/&gt;
 *       &lt;/sequence&gt;
 *     &lt;/restriction&gt;
 *   &lt;/complexContent&gt;
 * &lt;/complexType&gt;
 * </pre>
 *
 * <p>
 * <b>Thread safety</b>: This class is a bean, it's mutable, it's no threads safe, however it's not required because it
 * will be used in a single thread manner.
 * </p>
 *
 * @author TCSDESIGNER, TCSDEVELOPER
 * @version 1.0
 */
@XmlAccessorType(XmlAccessType.FIELD)
@XmlType(name = "prerequisiteDocument", propOrder = {"documentId", "version", "versionDate", "name", "content"})
public class PrerequisiteDocument implements Serializable {

    /**
     * Represents the serial version unique id.
     */
    private static final long serialVersionUID = 5765138167147465302L;

    /**
     * <p>
     * Represents the document id.
     * </p>
     */
    private long documentId = -1;

    /**
     * <p>
     * Represents the version.
     * </p>
     */
    private Integer version;

    /**
     * <p>
     * Represents the version date.
     * </p>
     */
    @XmlElement(required = true)
    private XMLGregorianCalendar versionDate;

    /**
     * <p>
     * Represents the name of the pre-requisite document.
     * </p>
     */
    @XmlElement(required = true)
    private String name;

    /**
     * <p>
     * Represents the content of the pre-requisite document.
     * </p>
     */
    @XmlElement(required = true)
    private String content;

    /**
     * <p>
     * Creates a <code>PrerequisiteDocument</code> instance.
     * </p>
     */
    public PrerequisiteDocument() {
    }

    /**
     * <p>
     * Gets the document id.
     * </p>
     *
     * @return the document id, never less than -1, default to -1.
     */
    public long getDocumentId() {
        return documentId;
    }

    /**
     * <p>
     * Sets the document id.
     * </p>
     *
     * @param documentId
     *            the documentId to set
     * @throws IllegalArgumentException
     *             if documentId &lt; -1
     */
    public void setDocumentId(long documentId) {
        if (documentId < -1) {
            throw new IllegalArgumentException("the [documentId] should not less than -1.");
        }

        this.documentId = documentId;
    }

    /**
     * <p>
     * Gets the version.
     * </p>
     *
     * @return the version, possibly null, but its value is never negative.
     */
    public Integer getVersion() {
        return version;
    }

    /**
     * <p>
     * Sets the version.
     * </p>
     *
     * @param version
     *            the version to set, can be null
     * @throws IllegalArgumentException
     *             if it's not null and it's less than 0
     */
    public void setVersion(Integer version) {
        if (null != version && version.intValue() < 0) {
            throw new IllegalArgumentException("The version can not be negative.");
        }

        this.version = version;
    }

    /**
     * <p>
     * Gets the version date.
     * </p>
     *
     * @return the version date, possibly null.
     */
    public XMLGregorianCalendar getVersionDate() {
        return versionDate;
    }

    /**
     * <p>
     * Sets the version date.
     * </p>
     *
     * @param versionDate
     *            the versionDate to set, can be null
     */
    public void setVersionDate(XMLGregorianCalendar versionDate) {
        this.versionDate = versionDate;

    }

    /**
     * <p>
     * Gets the name.
     * </p>
     *
     * @return the name, possibly null or empty.
     */
    public String getName() {
        return name;
    }

    /**
     * <p>
     * Sets the name.
     * </p>
     *
     * @param name
     *            the name to set, can be null, can be empty
     */
    public void setName(String name) {
        this.name = name;
    }

    /**
     * <p>
     * Gets the content.
     * </p>
     *
     * @return the content, possibly null or empty.
     */
    public String getContent() {
        return content;
    }

    /**
     * <p>
     * Sets the content.
     * </p>
     *
     * @param content
     *            the content to set, can be null, can be empty
     */
    public void setContent(String content) {
        this.content = content;
    }
}

File: widget_bridge/trunk/src/js/main/js/topcoder/widgets/bridge/PrerequisiteDocument.js
/**
 * Copyright (C) 2008 TopCoder Inc., All Rights Reserved.
 *
 * Widget Bridge 1.0
 */

/**
 * The constructor takes JSON object and assign value to each internal fields.
 *
 * @param {JSONObject} json JSON object with will be converted to PrerequisiteDocument
 *
 * @consturctor
 * @namespace js.topcoder.widgets.bridge
 * @class PrerequisiteDocument
 *
 * This class is a client side mapping for PrerequisiteDocument entity in service side. It also encapsulates
 * JSON object coversion.
 * Thread Safety:
 * This class is mutable and not thread-safe. But JavaScript execution model is mostly single-threaded.
 * And there is no thread-synchronization features in JavaScript, so it is not an issue here. Anyway, the
 * external code should be careful if several threads are used to communicate with this class.
 *
 * @author TCSDEVELOPER
 * @version 1.0
 */
js.topcoder.widgets.bridge.PrerequisiteDocument = function(json){
    /**
     * This variable represents documentID. It can not be undefiend or null or less than -1 and its initial value
     *  is -1 which means it is not set yet.
     *
     *
     * @private
     * @type number
     */
    var documentID = -1;


    /**
     * This variable represents document version number. It can not be undefiend or null or less  than -1 and its
     * initial value is -1 which emans it is not set yet.
     *
     * @private
     * @type number
     */
    var version = -1;


    /**
     * This variable represents version date string. It can not be undefined or null. It should be assigned
     * date string as "2008-05-01 09:00".
     *
     * @private
     * @type string
     */
    var versionDate = "";

    /**
     * This variable represents document name. It can not be undefined or null.
     *
     * @private
     * @type string
     */
    var name = "";

    /**
     * This variable represents document content. It can not be undefined or null.
     *
     * @private
     * @type string
     */
    var content = "";

    /**
     * Abbreviation for Helper class.
     */
    var Helper = js.topcoder.widgets.bridge.Helper;

    /**
     * Abbreviation for IllegalArgumentException class.
     */
    var IllegalArgumentException = js.topcoder.widgets.bridge.IllegalArgumentException;

    /**
     * A simple setter for documentID internal field.
     *
     * @param {number} newDocumentID the value of document id.
     * @throws IllegalArgumentException if the argument is null or undefined value or less than 0.
     */
    this.setDocumentID = function(newDocumentID){
        Helper.checkGreaterOrEqual(newDocumentID, "document id", 0);
        documentID = newDocumentID;
    }

    /**
     * A simple getter for documentID internal field.
     *
     * @return {number} documentID internal field.
     */
    this.getDocumentID = function(){
        return documentID;
    }

    /**
     * A simple setter for version internal field.
     *
     * @param {number} newVersion the value of new version.
     * @throws IllegalArgumentException if the argument is null or undefined value or less than 0.
     */
    this.setVersion = function(newVersion){
        Helper.checkGreaterOrEqual(newVersion, "version", 0);
        version = newVersion;
    }

    /**
     * A simple getter for version internal field.
     *
     * @return {number} version internal field.
     */
    this.getVersion = function(){
        return version;
    }

    /**
     * A simple setter for versionDate internal field.
     *
     * @param {string} newVersionDate the value of new version date string
     * @throws IllegalArgumentException if the argument is null or undefined value or not string type
     */
    this.setVersionDate = function(newVersionDate){
        Helper.checkDefined(newVersionDate, "verionDate", "string");
        versionDate = newVersionDate;
    }

    /**
     * A simple getter for versionDate internal field.
     *
     * @return {string} versionDate internal field.
     */
    this.getVersionDate = function(){
        return versionDate;
    }

    /**
     * A simple setter for name internal field.
     *
     * @param {string} newName the value of new name
     * @throws IllegalArgumentException if the argument is null or undefined value or not string type
     */
    this.setName = function(newName){
        Helper.checkDefined(newName, "name", "string");
        name = newName;
    }

    /**
     * A simple getter for name internal field.
     *
     * @return {string} name internal field.
     */
    this.getName = function(){
        return name;
    }


    /**
     * A simple setter for content internal field.
     *
     * @param {string} newContent the value of new content
     * @throws IllegalArgumentException if the argument is null or undefined value or not string type
     */
    this.setContent = function(newContent){
        Helper.checkDefined(newContent, "content", "string");
        content = newContent;
    }

    /**
     * A simple getter for content internal field.
     *
     * @return {string} content internal field.
     */
    this.getContent = function(){
        return content;
    }

    /**
     * Convert the current instance into JSON string.
     *
     * @return a JSON string representing current object
     */
    this.toJSON = function(){
        var result = "{";
        result += "\"documentID\":" + documentID;
        result += ",\"version\":" + version;
        result += ",\"versionDate\":\"" + Helper.escapeJSONValue(versionDate)+"\"";
        result += ",\"name\":\"" + Helper.escapeJSONValue(name)+"\"";
        result += ",\"content\":\"" + Helper.escapeJSONValue(content)+"\"";
		result += "}";
		return result;
    }

    //constructor code
    if (json != null) {
        this.setDocumentID(json.documentID);
        this.setVersion(json.version);
        this.setVersionDate(json.versionDate);
        this.setName(json.name);
        this.setContent(json.content);
    }
}
